# Table of contents
# 1 Importing libraries
# 2 Importing data files
# 3 Mixed type data in test frame
# 4 Finding missing values in df_prods
# 5 Finding and addressing duplicates in df_prods
# 6 Exporting dataframes
# 7 Task
## Question 2: Run the df.describe() function on your df_ords dataframe
## Question 3: Check for mixed-type data in your df_ords dataframe
## Question 4: If you find mixed-type data, fix it. 
## Question 5: Check for missing values in df_ords dataframe
## Question 6: Address the missing values using an appropriate method
## Question 7: Check for duplicate values in your df_ords data
## Question 8: Address the duplicates using an appropriate method
## Question 9: Export your final, cleaned df_prods and df_ords data as “.csv” files 

# 1 Importing libraries

In [4]:
import pandas as pd
import numpy as np
import os

# 2 Importing data files

In [5]:
path = r'C:\Users\Eva\Documents\Instacart Basket Analysis'

In [6]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [7]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# 3 Mixed type data in test frame

In [8]:
## Create a dataframe with mixed type data
df_test = pd.DataFrame()

In [9]:
df_test['mix'] = ['a', 'b', 1, True]

In [10]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [11]:
## Checking for mixed type
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)
    

mix


In [12]:
## Changing the data type to string
df_test['mix'] = df_test['mix'].astype('str')

# 4 Finding missing values in df_prods

In [13]:
## Finding the columns with missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [14]:
## Viewing missing values subset
df_nan = df_prods[df_prods['product_name'].isnull()==True]

In [15]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [16]:
## There are 16 missing values found (also see above). 

In [17]:
## Rows and columns of the dataframe with missing values
df_prods.shape

(49693, 5)

In [18]:
## Rows and columns of the dataframe without missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [19]:
df_prods_clean.shape

(49677, 5)

# 5 Finding and addressing duplicates in df_prods

In [20]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [21]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [22]:
## 5 duplicates are found.

In [23]:
## Rows and columns of the dataframe with duplicates
df_prods_clean.shape

(49677, 5)

In [24]:
# Dropping the duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [25]:
## Rows and columns of the dataframe without duplicates
df_prods_clean_no_dups.shape

(49672, 5)

# 6 Exporting dataframes

In [45]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'), index = False)

# 7 Task

## Question 2: Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.

In [27]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

### I can drop the "unnamed: 0" columns because it is unnecessary. Apart from that, everything seems to be okay.

In [28]:
## Drop unnecessary column
df_ords = df_ords.drop(columns = ['Unnamed: 0'])

## Question 3: Check for mixed-type data in your df_ords dataframe

In [29]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)
    

### There is no mixed type data.

## Question 4: If you find mixed-type data, fix it. 

### There is no mixed type data.

## Question 5: Check for missing values in df_ords dataframe

In [30]:
## Finding missing values
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [31]:
## Creating a subset of the missing values
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [32]:
## View missing values subset
df_ords_nan

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
11        2168274        2             1                   2   
26        1374495        3             1                   1   
39        3343014        4             1                   6   
45        2717275        5             1                   3   
...           ...      ...           ...                 ...   
3420930    969311   206205             1                   4   
3420934   3189322   206206             1                   3   
3421002   2166133   206207             1                   6   
3421019   2227043   206208             1                   1   
3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

In [33]:
## Viewing orders of user with user_id 1
df_ords[df_ords['user_id'] == 1]

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   1187899        1            11                   4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

In [34]:
## Viewing orders of user with user_id 2
df_ords[df_ords['user_id'] == 2]

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
11   2168274        2             1                   2                 11   
12   1501582        2             2                   5                 10   
13   1901567        2             3                   1                 10   
14    738281        2             4                   2                 10   
15   1673511        2             5                   3                 11   
16   1199898        2             6                   2                  9   
17   3194192        2             7                   2                 12   
18    788338        2             8                   1                 15   
19   1718559        2             9                   2                  9   
20   1447487        2            10                   1                 11   
21   1402090        2            11                   1                 10   
22   3186735        2            12                   1                  9   
23   3268552        2            13                   4                 11   
24    839880        2            14                   3                 10   
25   1492625        2            15                   1                 11   

    days_since_prior_order  
11                     NaN  
12                    10.0  
13                     3.0  
14                     8.0  
15                     8.0  
16                    13.0  
17                    14.0  
18                    27.0  
19                     8.0  
20                     6.0  
21                    30.0  
22                    28.0  
23                    30.0  
24                    13.0  
25                    30.0

In [35]:
## Viewing orders of user with user_id 3
df_ords[df_ords['user_id'] == 3]

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
26   1374495        3             1                   1                 14   
27    444309        3             2                   3                 19   
28   3002854        3             3                   3                 16   
29   2037211        3             4                   2                 18   
30   2710558        3             5                   0                 17   
31   1972919        3             6                   0                 16   
32   1839752        3             7                   0                 15   
33   3225766        3             8                   0                 17   
34   3160850        3             9                   0                 16   
35    676467        3            10                   3                 16   
36    521107        3            11                   0                 18   
37   1402502        3            12                   1                 15   
38   2774568        3            13                   5                 15   

    days_since_prior_order  
26                     NaN  
27                     9.0  
28                    21.0  
29                    20.0  
30                    12.0  
31                     7.0  
32                     7.0  
33                     7.0  
34                     7.0  
35                    17.0  
36                    11.0  
37                    15.0  
38                    11.0

### There are 206209 missing values in the column "days since prior order". As the examples from above show, missing values in this column mean that the customer ordered something for the very first time (order_number = 1), so no value could have been entered.

## Question 6: Address the missing values using an appropriate method

In [36]:
df_ords['days_since_prior_order'].fillna(0, inplace=True)

### I imputed 0 values for the NaN values, so that the values tell us that a customer has made its first order.

In [37]:
df_ords.head(30)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   1187899        1            11                   4                  8   
11   2168274        2             1                   2                 11   
12   1501582        2             2                   5                 10   
13   1901567        2             3                   1                 10   
14    738281        2             4                   2                 10   
15   1673511        2             5                   3                 11   
16   1199898        2             6                   2                  9   
17   3194192        2             7                   2                 12   
18    788338        2             8                   1                 15   
19   1718559        2             9                   2                  9   
20   1447487        2            10                   1                 11   
21   1402090        2            11                   1                 10   
22   3186735        2            12                   1                  9   
23   3268552        2            13                   4                 11   
24    839880        2            14                   3                 10   
25   1492625        2            15                   1                 11   
26   1374495        3             1                   1                 14   
27    444309        3             2                   3                 19   
28   3002854        3             3                   3                 16   
29   2037211        3             4                   2                 18   

    days_since_prior_order  
0                      0.0  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0  
11                     0.0  
12                    10.0  
13                     3.0  
14                     8.0  
15                     8.0  
16                    13.0  
17                    14.0  
18                    27.0  
19                     8.0  
20                     6.0  
21                    30.0  
22                    28.0  
23                    30.0  
24                    13.0  
25                    30.0  
26                     0.0  
27                     9.0  
28                    21.0  
29                    20.0

## Question 7: Check for duplicate values in your df_ords data

In [38]:
df_dups = df_ords[df_ords.duplicated()]

In [39]:
df_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

### No duplicate data was found.

## Question 8: Address the duplicates using an appropriate method

### No duplicate data was found, so I don't need to do anything.

## Question 9: Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [44]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'), index = False)